### PR TITLE
[xDS GKE] use randomized local forwarding port for parallism

### PR DESCRIPTION
**Pull request by lidizheng**
While implementing parallel run with Bazel in https://github.com/grpc/grpc/pull/27170, I found we have one mutual exclusive resource -- the local forwarding port. This technically won't impact CI, since they use direct connection, but it's always nice to be able to develop locally.

Tested with the CSDS test case. See test run log: https://gist.github.com/lidizheng/d839200a7d8781cb0c4420f8d922006a

### CODE REVIEW

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/client_app.py [hunk 1]
```diff
@@ -344,18 +345,20 @@ def run(self,
...
         pod_ip = pod.status.pod_ip
+        rpc_port = self.stats_port
         rpc_host = None
 
         # Experimental, for local debugging.
         if self.debug_use_port_forwarding:
             logger.info('LOCAL DEV MODE: Enabling port forwarding to %s:%s',
                         pod_ip, self.stats_port)
-            self.port_forwarder = self.k8s_namespace.port_forward_pod(
+            self.local_forwarding_port, self.port_forwarder = self.k8s_namespace.port_forward_pod(
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/test_app/client_app.py:355]: Minor fix needed: unset `self.local_forwarding_port` in `cleanup()` after `self.k8s_namespace.port_forward_stop(self.port_forwarder)`

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py [hunk 2]
```diff
@@ -325,25 +323,22 @@ def run(self,
...
-                    self.k8s_namespace.port_forward_pod(
-                        pod,
-                        remote_port=maintenance_port,
-                        local_port=local_port))
+                logger.info('LOCAL DEV MODE: Enabling port forwarding to %s:%s',
+                            pod_ip, maintenance_port)
+                local_port, port_forwarder = self.k8s_namespace.port_forward_pod(
+                    pod, remote_port=maintenance_port)
+                self.port_forwarders.append(port_forwarder)
+                self.local_forwarding_ports.append(local_port)
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py:335]: Same here: unset `self.local_forwarding_ports` in `self.cleanup()`

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py [hunk 3]
```diff
@@ -329,18 +335,32 @@ def port_forward_pod(
...
+
+                logger.info(output)
+                break
         except Exception:
             self.port_forward_stop(pf)
             raise
 
         # TODO(sergiitk): return new PortForwarder object
-        return pf
+        return local_port, pf
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:363]: Nit: maybe we could create a `PortForwarder` class that will wrap `subprocess.Popen` and the port?

**Review by sergiitk — APPROVED** [review]
Great idea, thank you! Left some minor comments.

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py (hunk 3 above)
Comment (sergiitk) [reply tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:363]: We also could move `k8s_namespace.port_forward_stop.port_forward_stop(pf):` static method to a concrete method of `PortForwarder`.

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/client_app.py (hunk 1 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/test_app/client_app.py]: `local_forwarding_port` removed, for the `PortForwarder` class.

File: tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py (hunk 2 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/test_app/server_app.py]: `local_forwarding_ports` removed, for the `PortForwarder` class.

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py [hunk 4]
```diff
@@ -78,9 +79,92 @@ class PortForwardingError(Exception):
...
+                 destination: str,
+                 remote_port: int,
+                 local_port: Optional[int] = None,
+                 local_address: Optional[str] = None):
+        self.context = context
+        self.namespace = namespace
+        self.destination = destination
+        self.remote_port = remote_port
+        self.local_address = local_address or self.PORT_FORWARD_LOCAL_ADDRESS
+        self.local_port, self.subprocess = self._establish_port_forwarding(
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:97]: Normally I'd not be doing any heavy work in the constructor, and have separate `connect` (so class can be instantiated for unit tests and such), but it doesn't matter as much in this case.

**Comment by lidizheng** [comment]
Python url map tests: [sponge/fd899a7f-ae90-465e-befd-1c4fb3f7cd1f](http://sponge/fd899a7f-ae90-465e-befd-1c4fb3f7cd1f)
C++ K8s LB tests: [sponge/f714dd97-a487-4beb-9346-792e11827d80](http://sponge/f714dd97-a487-4beb-9346-792e11827d80)

Manually validated with the csds test case.

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py [hunk 5]
```diff
@@ -78,9 +79,92 @@ class PortForwardingError(Exception):
...
+                    if groups is None:
+                        raise PortForwardingError(
+                            f'Error forwarding port, unexpected output {output}'
+                        )
+                    local_port = int(groups[1])
+
+                logger.info(output)
+                break
+        except Exception:
+            self.port_forward_stop(pf)
```
Comment (sergiitk) [inline tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py:154]:
```suggestion
            self.close()
```

You'd need to set `self.subprocess = pf` too

**Review by sergiitk — CHANGES_REQUESTED** [review]
`self.port_forward_stop()` will trigger an error

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py (hunk 4 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py]: Done.

File: tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py (hunk 5 above)
Comment (lidizheng) [reply tools/run_tests/xds_k8s_test_driver/framework/infrastructure/k8s.py]: Done, with the above `connect` change, this function now is a state mutating method.

**Review by sergiitk — APPROVED** [review]
Great! Thank you

**Pull request merged**